demo/image_creation.sh

```diff
@@ -1,125 +1,162 @@
+FOUND=0
+while [[ $(date '+%s') < $DEADLINE ]]
 do
-    power_state=$(az vm get-instance-view -g $rg -n $vm | jq '.instanceView.statuses[] | select(.code | contains("PowerState")) | .code' -r)
-    if [[ "$power_state" == *"stopped"* ]]
+    POWER_STATE=$(az vm get-instance-view -g "$RG" -n "$VM_NAME" | jq '.instanceView.statuses[] | select(.code | contains("PowerState")) | .code' -r)
+    if [[ "$POWER_STATE" == *"stopped"* ]]
     then
-        log=$(az vm boot-diagnostics get-boot-log -g $rg -n $vm)
-        if [[ -n "$log" && "$log" == *"SIGTOOL_END"* ]]
+        LOG=$(az vm boot-diagnostics get-boot-log -g "$RG" -n "$VM_NAME")
+        if [[ -n "$LOG" && "$LOG" == *"SIGTOOL_END"* ]]
         then
             echo "vm configured successfully"
-            found=1
+            FOUND=1
             break
         fi
     fi
     sleep 15
 done
 
-if [[ $found -eq 0 ]]
+if [[ $FOUND -eq 0 ]]
 then
     echo "vm failed to configure in 15 minutes - abort"
     exit 1
 fi
 
-image=image-$epoch
+IMAGE_NAME=image-$EPOCH
+echo "*********************************************************************"
+echo "Capturing OsDisk snapshot of vm $VM_NAME with image $IMAGE_NAME"
+echo "*********************************************************************"
+TARGET_DISK=$(az disk show --ids "$(az vm show -g "$RG" -n "$VM_NAME" | jq .storageProfile.osDisk.managedDisk.id -r)" | jq .name -r)
+az snapshot create -g "$RG" -n "$VM_NAME"-snapshot --source "$TARGET_DISK"
+
+IMAGE_VERSION=$(date '+%Y.%m%d.%H%M%S')
+GALLERY_NAME=testgalleryazinit
+GALLERY_DEFINITION=testgalleryazinitdef
+GALLERY_RG=testgalleryazinitrg
+SNAPSHOT_ID=$(az snapshot show -n "$VM_NAME"-snapshot -g "$RG" --query id --output tsv)
+
+# If BASE_IMAGE == canonical:0001-com-ubuntu-server-jammy:22_04-lts-gen2:latest,
+# GALLERY_PUBLISHER = canonical
+# GALLERY_OFFER = 0001-com-ubuntu-server-jammy
+# GALLERY_SKU = 22_04-lts-gen2
+GALLERY_PUBLISHER="$(echo "$BASE_IMAGE" | cut -f1 -d:)"
+GALLERY_OFFER="$(echo "$BASE_IMAGE" | cut -f2 -d:)"
+GALLERY_SKU="$(echo "$BASE_IMAGE" | cut -f3 -d:)"
+
+az group create --location "$LOCATION" --resource-group "$GALLERY_RG"
+
+az sig create --resource-group "$GALLERY_RG" --gallery-name "$GALLERY_NAME"
+
 echo "*********************************************************************"
-echo "Capturing OsDisk snapshot of vm $vm with image $image"
+echo "Creating image definition of $BASE_IMAGE"
 echo "*********************************************************************"
-target_disk=$(az disk show --ids $(az vm show -g $rg -n $vm | jq .storageProfile.osDisk.managedDisk.id -r) | jq .name -r)
-az snapshot create -g $rg -n $vm-snapshot --source $target_disk
+az sig image-definition create --resource-group "$GALLERY_RG" --gallery-name "$GALLERY_NAME" --gallery-image-definition "$GALLERY_DEFINITION" --offer "$GALLERY_OFFER" --publisher "$GALLERY_PUBLISHER" --sku "$GALLERY_SKU" --os-type linux --os-state Specialized --features SecurityType="$SECURITY_TYPE"
```

Comment:
os state for the image definition should be Generalized and not specialized
The reason is that specialized deployment in Azure does not wait or need provisioning complete signal. The deployment of specialized image also does not allow passing in user/hostname etc...

Reply:
Done